### PR TITLE
Aborting deployment.

### PR DESCRIPTION
**Pull request by kjaskiewiczz**
https://tracker.mender.io/browse/MEN-718
https://tracker.mender.io/browse/MEN-711

@bboozzoo @maciejmrowiec @mchalski

### CODE REVIEW

**Comment by bboozzoo** [comment]
Tests are failing

**Comment by kjaskiewiczz** [comment]
I know; I'm updating tests right now

File: resources/deployments/controller/controller_deployments.go [hunk 1]
```diff
@@ -127,6 +128,36 @@ func (d *DeploymentsController) GetDeploymentStats(w rest.ResponseWriter, r *res
...
+
+	if !govalidator.IsUUIDv4(id) {
+		d.view.RenderError(w, r, ErrIDNotUUIDv4, http.StatusBadRequest, l)
+		return
+	}
+
+	// Check if deployment is finished
+	isDeploymentFinished, err := d.model.IsDeploymentFinished(id)
+	if err != nil {
+		d.view.RenderInternalError(w, r, err, l)
```
Comment (bboozzoo) [inline resources/deployments/controller/controller_deployments.go]: Shouldn't we respond with 409 Conflict perhaps? Would require docs update as well.

Comment (bboozzoo) [inline resources/deployments/controller/controller_deployments.go]: My bad, missed that we're returning an error from DB layer here.

File: resources/deployments/controller/controller_deployments.go [hunk 2]
```diff
@@ -127,6 +128,36 @@ func (d *DeploymentsController) GetDeploymentStats(w rest.ResponseWriter, r *res
...
+		return
+	}
+	if isDeploymentFinished {
+		d.view.RenderError(w, r, ErrDeploymentAlreadyFinished, http.StatusUnprocessableEntity, l)
```
Comment (kjaskiewiczz) [inline resources/deployments/controller/controller_deployments.go]: done

File: resources/deployments/mongo/device_deployments.go [hunk 3]
```diff
@@ -38,6 +39,13 @@ const (
 	StorageKeyDeviceDeploymentStatus          = "status"
 	StorageKeyDeviceDeploymentDeploymentID    = "deploymentid"
 	StorageKeyDeviceDeploymentFinished        = "finished"
+
+	// device deployment statuses
```
Comment (kjaskiewiczz) [inline resources/deployments/mongo/device_deployments.go]: right, fixed

File: resources/deployments/mongo/device_deployments.go [hunk 4]
```diff
@@ -308,3 +316,34 @@ func (d *DeviceDeploymentsStorage) GetDeviceDeploymentStatus(deploymentID string
...
+
+func (d *DeviceDeploymentsStorage) AbortDeviceDeployments(deploymentId string) error {
+	if govalidator.IsNull(deploymentId) {
+		return ErrStorageInvalidID
+	}
+
+	session := d.session.Copy()
+	defer session.Close()
+	selector := bson.M{"$and": []bson.M{
+		bson.M{"deploymentid": deploymentId},
```
Comment (kjaskiewiczz) [inline resources/deployments/mongo/device_deployments.go]: done

File: resources/deployments/mongo/device_deployments.go [hunk 5]
```diff
@@ -308,3 +316,34 @@ func (d *DeviceDeploymentsStorage) GetDeviceDeploymentStatus(deploymentID string
...
+					DeviceDeploymentStatusInstalling,
+					DeviceDeploymentStatusRebooting,
+					DeviceDeploymentStatusPending,
+				}}},
+	}}
+
+	update := bson.M{"$set": bson.M{"status": DeviceDeploymentStatusAborted}}
+
+	info, err := session.DB(DatabaseName).C(CollectionDevices).UpdateAll(selector, update)
+	fmt.Println("Update info: ", info)
```
Comment (kjaskiewiczz) [inline resources/deployments/mongo/device_deployments.go]: fixed

File: resources/deployments/mongo/deployments.go [hunk 6]
```diff
@@ -160,6 +160,50 @@ func (d *DeploymentsStorage) FindByID(id string) (*deployments.Deployment, error
...
+
+	if govalidator.IsNull(id) {
+		return nil, ErrStorageInvalidID
+	}
+
+	session := d.session.Copy()
+	defer session.Close()
+
+	var deployment *deployments.Deployment
+	filter := bson.M{"_id": id, "finished": time.Time{}}
```
Comment (kjaskiewiczz) [inline resources/deployments/mongo/deployments.go]: done

**Pull request merged**